Bucketeer/Sources/Internal/Evaluation/EvaluationStorageImpl.swift

```diff
@@ -0,0 +1,110 @@
+        }
+        set {
+            evaluationUserDefaultsDao.evaluatedAt = newValue
+        }
+    }
+
+    var userAttributesUpdated: Bool {
+        get {
+            return evaluationUserDefaultsDao.userAttributesUpdated
+        }
+        set {
+            evaluationUserDefaultsDao.userAttributesUpdated = newValue
+        }
+    }
+
+    private let userId: String
+    // Expected SQL Dao
+    private let evaluationDao: EvaluationDao
+    // Expected in-memory cache Dao
+    private let evaluationMemCacheDao: EvaluationMemCacheDao
+    // using `var` for mutable evaluationUserDefaultDao properties | `let` will return error "could not mutable the immutable object"
+    private var evaluationUserDefaultsDao: EvaluationUserDefaultsDao
+
+    init(
+        userId: String,
+        evaluationDao: EvaluationDao,
+        evaluationMemCacheDao: EvaluationMemCacheDao,
+        evaluationUserDefaultsDao: EvaluationUserDefaultsDao
+    ) {
+        self.userId = userId
+        self.evaluationDao = evaluationDao
+        self.evaluationUserDefaultsDao = evaluationUserDefaultsDao
+        self.evaluationMemCacheDao = evaluationMemCacheDao
+        try? refreshCache()
+    }
+
+    func get(userId: String) throws -> [Evaluation] {
+        evaluationMemCacheDao.get(key: userId) ?? []
+    }
+
+    func deleteAllAndInsert(userId: String, evaluations: [Evaluation], evaluatedAt: String) throws {
+        try evaluationDao.deleteAll(userId: userId)
+        try evaluationDao.put(userId: userId, evaluations: evaluations)
+        // Update cache directly after we have force update
+        evaluationMemCacheDao.set(key: userId, value: evaluations)
+        self.evaluatedAt = evaluatedAt
+    }
+
+    func update(evaluations: [Evaluation], archivedFeatureIds: [String], evaluatedAt: String) throws -> Bool {
+        // 1. Update evaluation with new data
+        try evaluationDao.put(userId: userId, evaluations: evaluations)
+        // 2. Get current data in db
+        let evaluationsInDb = try evaluationDao.get(userId: userId)
+        var archivedEvaluations = [Evaluation]()
+        var activeEvaluations = [Evaluation]()
+        // 3. Filter active & archived
+        for evaluation in evaluationsInDb {
+            if archivedFeatureIds.contains(evaluation.featureId) {
+                archivedEvaluations.append(evaluation)
+            } else {
+                activeEvaluations.append(evaluation)
+            }
+        }
+        let archivedEvaluationIds = archivedEvaluations.map { $0.id }
+        // 4. Remove all evaluation which has `id` in the list `archivedEvaluationIds`
+        if (archivedEvaluationIds.count > 0) {
+            try evaluationDao.deleteByIds(archivedEvaluationIds)
+        }
+        self.evaluatedAt = evaluatedAt
+        // 5. Save a new cache
+        evaluationMemCacheDao.set(key: userId, value: activeEvaluations)
+
+        return evaluations.count > 0 || archivedEvaluationIds.count > 0
```

Comment:
We should wrap DB-related operations with a transaction.

Reply:
fixed